lib/spack/spack/cmd/external.py

```diff
@@ -0,0 +1,144 @@
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from __future__ import print_function
+from collections import defaultdict
+import os
+import re
+
+import spack
+import llnl.util.tty as tty
+
+description = "add external packages to Spack configuration"
+section = "config"
+level = "short"
+
+
+def setup_parser(subparser):
+    sp = subparser.add_subparsers(
+        metavar='SUBCOMMAND', dest='external_command')
+
+    # Find
+    sp.add_parser('find', help=external_find.__doc__)
+
+
+def _get_system_executables():
+    path = os.getenv('PATH')
+    search_paths = path.split(os.pathsep)
+    exe_to_path = {}
+    # Reverse order of search directories so that an exe in the first PATH
+    # entry overrides later entries
+    for search_path in reversed(search_paths):
+        for exe in os.listdir(search_path):
+            exe_to_path[exe] = os.path.join(search_path, exe)
+    return exe_to_path
+
+
+def _pkg_yaml_template(name, spec, path):
+    template = """\
+{name}:
+  paths:
+    {spec}: {path}"""
+    return template.format(name=name, spec=spec, path=path)
+
+
+def external_find(args):
+    _get_external_packages(TestRepo())
+
+
+def _get_external_packages(repo):
+    system_exe_to_path = _get_system_executables()
+
+    exe_pattern_to_pkgs = defaultdict(list)
+    for pkg in repo.all_packages():
+        if hasattr(pkg, 'executables'):
+            for exe in pkg.executables:
+                exe_pattern_to_pkgs[exe].append(pkg)
+
+    pkg_to_found_exes = defaultdict(set)
+    found_exe_to_pkgs = defaultdict(set)
+    for exe in system_exe_to_path:
+        for exe_pattern, pkgs in exe_pattern_to_pkgs.items():
+            if re.search(exe_pattern, exe):
+                for pkg in pkgs:
+                    pkg_to_found_exes[pkg].add(exe)
+                found_exe_to_pkgs[exe].update(pkgs)
+    # Sort to get repeatable results
+    found_exe_to_pkgs = dict(
+        (x, list(sorted(y, key=lambda p: p.name)))
+        for x, y in found_exe_to_pkgs.items())
+    pkg_to_found_exes = dict(
+        (x, list(sorted(y)))
+        for x, y in pkg_to_found_exes.items())
+
+    exe_to_usable_packages = defaultdict(list)
+    pkg_to_template = {}
+    resolved_pkgs = set()
+    for exe in sorted(found_exe_to_pkgs):
+        associated_packages = found_exe_to_pkgs[exe]
+        for pkg in associated_packages:
+            if pkg.name in resolved_pkgs:
+                # The package was already considered as part of examining a
+                # different exe. It does not need to be reexamined.
+                continue
+            resolved_pkgs.add(pkg.name)
+
+            found_pkg_exes = pkg_to_found_exes[pkg]
+
+            if hasattr(pkg, 'determine_spec_details'):
+                spec_str = pkg.determine_spec_details(found_pkg_exes)
+            else:
+                spec_str = pkg.name
```

Comment:
Does this assume a single spec? For compilers we might be dealing with cases like:
```console
$ gcc<Tab>
gcc             gcc-7           gcc-ar-4.8      gcc-ar-8        gcc-nm-5        gcc-nm-9        gcc-ranlib-6    
gcc-4.8         gcc-8           gcc-ar-5        gcc-ar-9        gcc-nm-6        gcc-ranlib      gcc-ranlib-7    
gcc-5           gcc-9           gcc-ar-6        gcc-nm          gcc-nm-7        gcc-ranlib-4.8  gcc-ranlib-8    
gcc-6           gcc-ar          gcc-ar-7        gcc-nm-4.8      gcc-nm-8        gcc-ranlib-5    gcc-ranlib-9   
```
so I'm wondering if that is covered here. What I would expect above is that multiple specs are deduced from the same package.

Reply:
This part doesn't but the logic above this needs to be adjusted

```
            if pkg.name in resolved_pkgs:
                 # The package was already considered as part of examining a
                 # different exe. It does not need to be reexamined.
                 continue
             resolved_pkgs.add(pkg.name)
```

it assumes we'll find one instance of any given package.